File: starters/cloud-chat/sync.py
#!/usr/bin/env python3
"""Terminal history sync: capture cursor before snapshot; refetch on change/reset."""
import argparse
import json
from pathlib import Path
import sys
sys.path.insert(0, str(Path(__file__).resolve().parents[2] / 'scripts/cloud'))
from bog_client import Client, Failure


def snapshots(client, limit=100):
    cursor = client.wait(timeout=0)['cursor']
    yield client.top('recent', limit=limit, include_fields=['/role', '/content', '/created_at'])
    while True:
        change = client.wait(cursor)
        cursor = change['cursor']
        if change['changed'] or change['reset']:
            yield client.top('recent', limit=limit, include_fields=['/role', '/content', '/created_at'])


if __name__ == '__main__':
    p=argparse.ArgumentParser();p.add_argument('--config',required=True);a=p.parse_args()
    try:
        for history in snapshots(Client.from_config(a.config)): print(json.dumps(history),flush=True)
    except KeyboardInterrupt: pass
    except (Failure,OSError,ValueError):
        print('History sync stopped; check connection and private configuration.',file=sys.stderr);sys.exit(1)

File: starters/cloud-chat/sync.mjs
import {Client} from '../../clients/typescript/src/index.js';
export async function* snapshots(client,limit=100){
  let cursor=(await client.wait(undefined,0)).cursor;
  yield await client.top('recent',limit,0,['/role','/content','/created_at']);
  while(true){const change=await client.wait(cursor);cursor=change.cursor;if(change.changed||change.reset)yield await client.top('recent',limit,0,['/role','/content','/created_at']);}
}
if(process.argv[1]===new URL(import.meta.url).pathname){try{const client=await Client.fromConfig(process.argv[2]);for await(const history of snapshots(client))console.log(JSON.stringify(history));}catch{console.error('History sync stopped; check connection and private configuration.');process.exitCode=1;}}
